syntax = "proto3";
import "google/protobuf/empty.proto";

service OobleckMaster {
    rpc GetDistInfo(google.protobuf.Empty) returns (DistInfo) {}
    rpc GetCode(google.protobuf.Empty) returns (CodeInfo) {}
    rpc SendMasterRankPort(PortInfo) returns (google.protobuf.Empty) {}
    rpc ReceiveMasterRankPort(google.protobuf.Empty) returns (PortInfo) {}
    rpc ReceiveReconfigurationNotification(google.protobuf.Empty) returns (stream DistInfo) {}
}

message HostInfo {
    string ip = 1;
    uint32 slots = 2;
    uint32 port = 3;
}

message DistInfo {
    repeated HostInfo hosts = 1;
}

message CodeInfo {
    bytes code = 1;
}

message PortInfo {
    uint32 port = 1;
}